src/utils/metrics.rs

```diff
@@ -0,0 +1,167 @@
+    pub statsd_client: StatsdClient,
+    prefix: String,
+}
+
+impl Metrics for MetricsClient {
+    fn counter(&self, key: &str, value: Option<i64>, tags: Option<HashMap<&str, &str>>) {
+        let mut count_value: i64 = 1;
+        if let Some(value) = value {
+            count_value = value;
+        }
+
+        if let Some(tags) = tags {
+            let mut metric_builder = self.statsd_client.count_with_tags(key, count_value);
+            for (key, value) in tags {
+                metric_builder = metric_builder.with_tag(key, value);
+            }
+            let result = metric_builder.try_send();
+            match result {
+                Ok(_) => {}
+                Err(_err) => {
+                    println!("Failed to send metric {}: {}", key, _err)
+                }
+            }
+        } else {
+            let result = self.statsd_client.count(key, count_value);
+            match result {
+                Ok(_) => {}
+                Err(_err) => {
+                    println!("Failed to send metric {}: {}", key, _err)
+                }
+            }
+        }
+    }
+
+    fn gauge(&self, key: &str, value: u64, tags: Option<HashMap<&str, &str>>) {
+        if let Some(tags) = tags {
+            let mut metric_builder = self.statsd_client.gauge_with_tags(key, value);
+            for (key, value) in tags {
+                metric_builder = metric_builder.with_tag(key, value);
+            }
+            let result = metric_builder.try_send();
+            match result {
+                Ok(_) => {}
+                Err(_err) => {
+                    println!("Failed to send metric {}: {}", key, _err)
+                }
+            }
+        } else {
+            let result = self.statsd_client.gauge(key, value);
+            match result {
+                Ok(_) => {}
+                Err(_err) => {
+                    println!("Failed to send metric {}: {}", key, _err)
+                }
+            }
+        }
+    }
+
+    fn time(&self, key: &str, value: u64, tags: Option<HashMap<&str, &str>>) {
+        if let Some(tags) = tags {
+            let mut metric_builder = self.statsd_client.time_with_tags(key, value);
+            for (key, value) in tags {
+                metric_builder = metric_builder.with_tag(key, value);
+            }
+            let result = metric_builder.try_send();
+            match result {
+                Ok(_) => {}
+                Err(_err) => {
+                    println!("Failed to send metric {}: {}", key, _err)
+                }
+            }
+        } else {
+            let result = self.statsd_client.time(key, value);
+            match result {
+                Ok(_) => {}
+                Err(_err) => {
+                    println!("Failed to send metric {}: {}", key, _err)
+                }
+            }
+        }
+    }
+}
+
+lazy_static! {
+    static ref METRICS_CLIENT: RwLock<Option<Arc<MetricsClient>>> = RwLock::new(None);
```

Comment:
How would this work if we start doing multi threading ? I guess there is where the thread local version may be helpful.

Reply:
In multi threading environment, we would use https://docs.rs/thread_local/latest/thread_local/. All the threads would get a read copy of this static variable. We use a rwlock since the workload is mostly readonly. We only need to write during initialization.